packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesAppLoaderTask.m

```diff
@@ -0,0 +1,252 @@
+    _selectionPolicy = selectionPolicy;
+    _delegateQueue = delegateQueue;
+    _loaderTaskQueue = dispatch_queue_create("expo.loader.LoaderTaskQueue", DISPATCH_QUEUE_SERIAL);
+  }
+  return self;
+}
+
+- (void)start
+{
+  if (!_config.isEnabled) {
+    dispatch_async(_delegateQueue, ^{
+      [self->_delegate appLoaderTask:self
+                  didFinishWithError:[NSError errorWithDomain:kEXUpdatesAppLoaderTaskErrorDomain code:1030 userInfo:@{
+                    NSLocalizedDescriptionKey: @"EXUpdatesAppLoaderTask was passed a configuration object with updates disabled. You should load updates from an embedded source rather than calling EXUpdatesAppLoaderTask, or enable updates in the configuration."
+                  }]];
+    });
+    return;
+  }
+
+  if (!_config.updateUrl) {
+    dispatch_async(_delegateQueue, ^{
+      [self->_delegate appLoaderTask:self
+                  didFinishWithError:[NSError errorWithDomain:kEXUpdatesAppLoaderTaskErrorDomain code:1030 userInfo:@{
+                    NSLocalizedDescriptionKey: @"EXUpdatesAppLoaderTask was passed a configuration object with a null URL. You must pass a nonnull URL in order to use EXUpdatesAppLoaderTask to load updates."
+                  }]];
+    });
+    return;
+  }
+
+  if (!_directory) {
+    dispatch_async(_delegateQueue, ^{
+      [self->_delegate appLoaderTask:self
+                  didFinishWithError:[NSError errorWithDomain:kEXUpdatesAppLoaderTaskErrorDomain code:1030 userInfo:@{
+                    NSLocalizedDescriptionKey: @"EXUpdatesAppLoaderTask directory must be nonnull."
+                  }]];
+    });
+    return;
+  }
+
+  BOOL shouldCheckForUpdate = [EXUpdatesUtils shouldCheckForUpdateWithConfig:_config];
+  NSNumber *launchWaitMs = _config.launchWaitMs;
+  if ([launchWaitMs isEqualToNumber:@(0)] || !shouldCheckForUpdate) {
+    self->_isTimerFinished = YES;
+  } else {
+    NSDate *fireDate = [NSDate dateWithTimeIntervalSinceNow:[launchWaitMs doubleValue] / 1000];
+    self->_timer = [[NSTimer alloc] initWithFireDate:fireDate interval:0 target:self selector:@selector(_timerDidFire) userInfo:nil repeats:NO];
+    [[NSRunLoop mainRunLoop] addTimer:self->_timer forMode:NSDefaultRunLoopMode];
+  }
+
+  [self _loadEmbeddedUpdateWithCompletion:^{
+    [self _launchWithCompletion:^(NSError * _Nullable error, BOOL success) {
+      if (!success) {
+        if (!shouldCheckForUpdate){
+          [self _finishWithError:error];
+        }
+        NSLog(@"Failed to launch embedded or launchable update: %@", error.localizedDescription);
+      } else {
+        self->_isReadyToLaunch = YES;
+        [self _maybeFinish];
+      }
+
+      if (shouldCheckForUpdate) {
+        [self _loadRemoteUpdateWithCompletion:^(NSError * _Nullable error, EXUpdatesUpdate * _Nullable update) {
+          [self _handleRemoteUpdateLoaded:update error:error];
+        }];
+      }
+    }];
+  }];
+}
+
+- (void)_finishWithError:(nullable NSError *)error
+{
```

Comment:
maybe assert that _isTimerFinished is set when this is called

Reply:
we call this method when we want to finish no matter whether the timer is finished or not (e.g. if there's an error loading an update from disk, but the configuration is set to not download any remote updates -- in that case we want to fail right away rather than waiting for the timer).

Reply:
Should we clear the timer in that case?

Reply:
yes, we do that at the end of the method: https://github.com/expo/expo/pull/9144/files#diff-c96dfcdb85e622163f90e2cdf6830020R141-R143 . do you think it should move earlier?

Reply:
maybe the right thing is to assert that this method always runs on the `loaderTaskQueue`.

Reply:
The dispatch_assert_queue looks good. These code paths don't run a lot, the cost of the extra assertion will be trivial.